### PR TITLE
Emulate AssignProjectConfiguration behavior in graph construction

### PR DESCRIPTION
**Pull request by dfederm**
### Summary
The for sln-based builds, the `AssignProjectConfiguration` task ends up using the Configuration and Platform defined in the sln rather than passing through the global properties from the referencing project or attempting to do dynamic platform negotiation. This change adds equivalent functionality to graph construction.

A concrete scenario this fixes for graph-based builds using an sln file is that most csproj define the "x86" platform while most vcxproj define "Win32". Previously for a graph build, if the csproj referenced the vcxproj, the platform passed to vcxproj would be x86, not Win32. Even worse, the vcxproj would be an entry point anyway, so it would double-build with both x86 AND Win32, which leads to race conditions.

### Customer Impact
Microsoft-internal customer using sln-based builds will be able to opt-into graph builds

### Regression?
No

### Testing
Manual validation in the customer repo, as well as added unit tests

### Risk
Low. Graph builds are a less-used feature, and this adds parity to what non-graph builds and VS-based builds do. It's unlikely that any behavioral change would be impactful due to those other scenarios presumably working for customers who may be using graph builds.

### CODE REVIEW

File: src/Build/Graph/ProjectInterpretation.cs [hunk 1]
```diff
@@ -118,22 +119,60 @@ public IEnumerable<ReferenceInfo> GetReferences(ProjectInstance requesterInstanc
...
 
-                var projectReferenceFullPath = projectReferenceItem.GetMetadataValue(FullPathMetadataName);
+                string projectReferenceFullPath = projectReferenceItem.GetMetadataValue(FullPathMetadataName);
+                bool enableDynamicPlatformResolution = ConversionUtilities.ValidBooleanTrue(requesterInstance.GetPropertyValue(EnableDynamicPlatformResolutionMetadataName));
+
+                PropertyDictionary<ProjectPropertyInstance> referenceGlobalProperties = GetGlobalPropertiesForItem(
+                    projectReferenceItem,
+                    requesterInstance.GlobalPropertiesDictionary,
+                    // Only allow reuse in scenarios where we will not mutate the collection.
+                    // TODO: Should these mutations be moved to globalPropertiesModifiers in the future?
```
Comment (dfederm) [inline src/Build/Graph/ProjectInterpretation.cs:129]: Note: I'm not planning on addressing this TODO in this PR. Dynamic platform resolution already deviated from this pattern, so a future PR should fix both of these.

File: src/Build/Graph/ProjectInterpretation.cs [hunk 2]
```diff
@@ -118,22 +119,55 @@ public IEnumerable<ReferenceInfo> GetReferences(ProjectInstance requesterInstanc
...
 
-                var referenceGlobalProperties = GetGlobalPropertiesForItem(projectReferenceItem, requesterInstance.GlobalPropertiesDictionary, ConversionUtilities.ValidBooleanTrue(requesterInstance.GetPropertyValue(EnableDynamicPlatformResolutionMetadataName)), globalPropertiesModifiers);
+                // Match what AssignProjectConfiguration does to resolve project references.
+                if (solutionConfiguration != null)
+                {
+                    string projectGuid = projectReferenceItem.GetMetadataValue(ProjectMetadataName);
+                    if (solutionConfiguration.TryGetProjectByGuid(projectGuid, out XmlElement projectElement)
+                        || solutionConfiguration.TryGetProjectByAbsolutePath(projectReferenceFullPath, out projectElement))
+                    {
+                        // Note: AssignProjectConfiguration sets FullConfiguration, SetConfiguration, and SetPlatform on the item, but these do not translate to global properties.
```
[the remark this answers is not in the document]
Comment (dfederm) [reply src/Build/Graph/ProjectInterpretation.cs:140]: `Configuration` and `Platform` become global props, but not `SetConfiguration` and `SetPlatform`.

[the remark this answers is not in the document]
Comment (dfederm) [reply src/Build/Graph/ProjectInterpretation.cs:140]: Basically this is describing how the behavior differs from `AssignProjectConfiguration`. Specifically if you look at `AssignProjectConfiguration.ResolveProject`, you'll see it sets `FullConfiguration`, `SetConfiguration`, `Configuration`, `SetPlatform`, and `Platform` as metadata. Only some of those item metadata actually translate to global properties in the evental `<MSBuild>` task call.

[the remark this answers is not in the document]
Comment (dfederm) [reply src/Build/Graph/ProjectInterpretation.cs:140]: I suppose actually the `SetConfiguration` item metadata -> `Configuration` global prop, and same with platform. So I guess it's actually the `FullConfiguration`, `Configuration`, and `Platform` item metadata which are ignored here. I dunno it's kinda confusing either way. Open to suggestion on how to clarify this comment.

File: src/Build/Graph/GraphBuilder.cs [hunk 3]
```diff
@@ -272,9 +272,12 @@ private static void AddEdgesFromSolution(IReadOnlyDictionary<ConfigurationMetada
...
             }
 
-            var projectsInSolution = GetBuildableProjects(solution);
+            IReadOnlyCollection<ProjectInSolution> projectsInSolution = GetBuildableProjects(solution);
 
-            var currentSolutionConfiguration = SelectSolutionConfiguration(solution, solutionGlobalProperties);
+            SolutionConfigurationInSolution currentSolutionConfiguration = SelectSolutionConfiguration(solution, solutionGlobalProperties);
+
+            string solutionConfigurationXml = SolutionProjectGenerator.GetSolutionConfiguration(solution, currentSolutionConfiguration);
+            solutionGlobalProperties = solutionGlobalProperties.SetItem("CurrentSolutionConfigurationContents", solutionConfigurationXml);
```
[the remark this answers is not in the document]
Comment (dfederm) [reply src/Build/Graph/GraphBuilder.cs:280]: Oh that's easy enough to do

File: src/Build/Graph/ProjectInterpretation.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (dfederm) [reply src/Build/Graph/ProjectInterpretation.cs]: Attempted to clarify, but open to further suggestion

**Review by rainersigwald — APPROVED** [review]

File: src/Build/Graph/ProjectInterpretation.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (rainersigwald) [reply src/Build/Graph/ProjectInterpretation.cs]: That works for me!

**Pull request merged**